Remove Magic Link Generation

rynowak opened the pull request:
This change resolves #3512 and #3636 by removing 'magic' link generation
and adding an extension method to add routes to areas correctly using the new
pattern. This is pretty much exactly the same as how MapWebApiRoute works.

For site authors, we recommend adding area-specific routes in a way that
includes a default AND constraint for the area. Put your most specific
(for link generation) routes FIRST.

Ex:
```
  routes.MapRoute(
     "admin_route"
      "Admin/{controller}/{action}/{id?}",
      defaults: new { area = "Admin" },
      constraints: new { area = "Admin" });
```

We're providing a new extension method that does this for you.
```
  routes.MapAreaRoute("admin_route", "Admin", "Admin/{controller}/{action}/{id?}");
```

The bulk of the changes here are to tests that unwittingly relied on the
old behavior.

rynowak commented on src/Microsoft.AspNet.Mvc.Core/Infrastructure/MvcRouteHandler.cs:
```diff
@@ -32,11 +32,7 @@ public VirtualPathData GetVirtualPath(VirtualPathContext context)
-            // The contract of this method is to check that the values coming in from the route are valid;
-            // that they match an existing action, setting IsBound = true if the values are OK.
-            context.IsBound = _actionSelector.HasValidAction(context);
+            context.IsBound = true;
```
This feature will be removed wholesale in a future routing change.

rynowak commented on test/Microsoft.AspNet.Mvc.FunctionalTests/InlineConstraintTests.cs:
```diff
@@ -408,7 +408,8 @@ public static IEnumerable<object[]> QueryParameters
                         "id",
-                        "sdsd", ""
+                        "sdsd",
+                        "/area-exists/InlineConstraints_Products/GetProductById?id=sdsd"
```
This route is now "greedy". I don't see much reason to try and rationalize this behavior, these tests still test what they did before - that the constraint in question doesn't allow that route to match with invalid values.

rynowak commented on test/Microsoft.AspNet.Mvc.FunctionalTests/RoutingTests.cs:
```diff
@@ -903,7 +903,7 @@ public RoutingTests(MvcTestFixture<RoutingWebSite.Startup> fixture)
 
         [Fact]
-        public async Task ConventionalRoutedAction_InArea_ImplicitLeaveArea()
+        public async Task ConventionalRoutedAction_InArea_StaysInArea()
```
Intentional behavior change.

rynowak commented on test/Microsoft.AspNet.Mvc.FunctionalTests/RoutingTests.cs:
```diff
@@ -985,7 +985,7 @@ public RoutingTests(MvcTestFixture<RoutingWebSite.Startup> fixture)
 
         [Fact]
-        public async Task AttributeRoutedAction_InArea_ImplicitLeaveArea()
+        public async Task AttributeRoutedAction_InArea_StaysInArea_ActionDoesntExist()
```
Intentional behavior change.

Eilon commented on src/Microsoft.AspNet.Mvc.Core/Infrastructure/MvcRouteHandler.cs:
```diff
@@ -32,11 +32,7 @@ public VirtualPathData GetVirtualPath(VirtualPathContext context)
                 throw new ArgumentNullException(nameof(context));
             }
 
-            EnsureServices(context.Context);
```
Is this gone because this was needed for action selection, which is no longer used?

rynowak replied:
Correct.

Eilon commented on test/Microsoft.AspNet.Mvc.FunctionalTests/compiler/resources/HtmlGenerationWebSite.HtmlGeneration_Home.Index.Encoded.html:
```diff
@@ -29,7 +29,7 @@
     </div>
     <div>
-        <a href="">
+        <a href="HtmlEncode[[unkonwn://localhost/HtmlGeneration_Product/List]]">
```
Is this a konwn unkonwn or an unkonwn unknown?

rynowak replied:
You go to war with the unkonwn you have

Eilon commented on test/WebSites/HtmlGenerationWebSite/Views/HtmlGeneration_Home/Index.cshtml:
```diff
@@ -35,7 +35,7 @@
     </div>
     <div>
-        <a asp-controller="Product" asp-protocol="unkonwn" asp-action="List">
+        <a asp-controller="HtmlGeneration_Product" asp-protocol="unkonwn" asp-action="List">
```
Also unknown.

dougbu replied:
The source of all that's unkonwn.

Eilon commented on src/Microsoft.AspNet.Mvc.Core/Builder/RouteBuilderExtensions.cs:
```diff
@@ -0,0 +1,136 @@
+    /// <summary>
+    /// Extension methods for <see cref="IRouteBuilder"/>.
+    /// </summary>
+    public static class RouteBuilderExtensions
```
Would like to rename this to something like `MvcAreaRouteBuilderExtensions` so it's pretty specific (because it's in a generic namespace). Or can remove `Area` from the name if we might add more non-area stuff here.

rynowak replied:
sure

dougbu replied:
Also suggest using these methods in the functional test sites or unit tests. Ignored in all tests at the moment.

Eilon commented:
:shipit: but yeah some of those tests look wonky to me. I agree it's fine to leave because they do capture some interesting behaviors.

rynowak commented:
> some of those tests look wonky to me

A lot of these tests really only touch this feature by accident. That's the trouble with these big html-generation/baseline style tests. We ran into similar issues when we started fixing bugs in modelbinding. Tests for unrelated features include bad behavior and verified bad data :+1:

Eilon commented:
Yup.

dougbu commented on src/Microsoft.AspNet.Mvc.Core/Builder/RouteBuilderExtensions.cs:
```diff
@@ -0,0 +1,136 @@
+    {
+        /// <summary>
+        /// Adds a route to the <see cref="IRouteBuilder"/> with the given MVC area with the specified
+        /// name and template.
```
Suggest:
- The `<summary>`s should vary to list all that's specified.
- Should use `<paramref>`.

So here `... specified <paramref name="name"/>, <paramref name="areaName"/> and <paramref name="template">.`

Separately did you intend to create an overload that only had `areaName` and `template` parameters i.e. used a default route `name`?

rynowak replied:
> Separately did you intend to create an overload that only had areaName and template parameters i.e. used a default route name?

No, `MapRoute` always has a positional parameter for route name, even though it's not required.

rynowak replied:
Added these details to the docs. Agreed that it seems dumb to list name, areaName and template, but not other things.

dougbu commented on src/Microsoft.AspNet.Mvc.Core/Builder/RouteBuilderExtensions.cs:
```diff
@@ -0,0 +1,136 @@
+                constraintsDictionary.Add("area", areaName);
+            }
+
+            return routeBuilder;
```
Something is missing here: Not using the `routeBuilder`, created dictionaries, or the remaining parameters.

rynowak replied:
:+1:

dougbu commented on src/Microsoft.AspNet.Mvc.Core/Builder/RouteBuilderExtensions.cs:
```diff
@@ -0,0 +1,136 @@
+                throw new ArgumentNullException(nameof(routeBuilder));
+            }
+
+            var defaultsDictionary = new RouteValueDictionary(defaults);
```
Why no `null` or `string.IsNullOrEmpty()` check of `name`?

rynowak replied:
It's allowed to be null or empty.

rynowak replied:
We should check `areaName` though.

dougbu commented on src/Microsoft.AspNet.Mvc.Core/Builder/RouteBuilderExtensions.cs:
```diff
@@ -0,0 +1,136 @@
+        /// </summary>
+        /// <param name="routeBuilder">The <see cref="IRouteBuilder"/> to add the route to.</param>
+        /// <param name="name">The name of the route.</param>
+        /// <param name="areaName">The MVC area name.</param>
```
I seem to remember `null` and empty area names have specific meanings. If so, would be useful to mention the semantics in these descriptions.

rynowak replied:
It only has a meaning in link generation. You can't define a null or empty area.

dougbu commented on test/Microsoft.AspNet.Mvc.FunctionalTests/compiler/resources/HtmlGenerationWebSite.HtmlGeneration_Home.Index.Encoded.html:
```diff
@@ -39,32 +39,32 @@
     </div>
     <div>
-        <a href="">
+        <a href="HtmlEncode[[unknown://localhost/NoControll#fragment]]">
```
Disappointed about the inconsistency in the unkonwn you have.

dougbu commented:
:watch:

rynowak commented:
Updated and added some unit tests. There aren't any good places I could find to use these in functional tests, because our tests all use the 'new' style `{area}/{controller}...` route.

dougbu commented on test/Microsoft.AspNet.Mvc.Core.Test/Builder/MvcAreaRouteBuilderExtensionsTest.cs:
```diff
@@ -0,0 +1,258 @@
+    public class MvcAreaRouteBuilderExtensionsTest
+    {
+        [Fact]
+        public void MapAreaRoute_Simple()
```
Shouldn't the test name say something about expectations e.g. `MapAreaRoute_Simple_DoesTheNeedful`?

rynowak replied:
> if you have nothing to say say nothing

All of these tests verify the most obvious possible thing. I don't see any reason to make the names more verbose.

dougbu commented:
:shipit: after correcting your spelling.
